packages/cursorless-engine/src/actions/JoinLines.ts

```diff
@@ -42,28 +56,59 @@ export default class JoinLines {
   }
 }
 
-function getEdits(editor: TextEditor, contentRanges: Range[]): Edit[] {
-  const { document } = editor;
+function getEdits(editor: TextEditor, targets: Target[]): Edit[] {
   const edits: Edit[] = [];
 
-  for (const range of contentRanges) {
-    const startLine = range.start.line;
-    const endLine = range.isSingleLine ? startLine + 1 : range.end.line;
+  for (const target of targets) {
+    const targetsEdits = target.joinAsLine
+      ? getLineTargetEditEdits(target)
+      : getTokenTargetEditEdits(target);
 
-    const lineIter = map(iterRange(startLine, endLine + 1), (i) =>
-      document.lineAt(i),
-    );
-    for (const [line1, line2] of pairwise(lineIter)) {
-      edits.push({
-        range: new Range(
-          line1.rangeTrimmed?.end ?? line1.range.end,
-          line2.rangeTrimmed?.start ?? line2.range.start,
-        ),
-        text: line2.isEmptyOrWhitespace ? "" : " ",
-        isReplace: true,
-      });
-    }
+    edits.push(...targetsEdits);
   }
 
   return edits;
 }
+
+function getTokenTargetEditEdits(target: Target): Edit[] {
+  const { editor, contentRange } = target;
+  const regex = getMatcher(editor.document.languageId).tokenMatcher;
+  const matches = generateMatchesInRange(
+    regex,
+    editor,
+    contentRange,
+    "forward",
+  );
+
+  return Array.from(pairwise(matches)).map(
+    ([range1, range2]): Edit => ({
+      range: new Range(range1.end, range2.start),
+      text: "",
+      isReplace: true,
+    }),
+  );
```

Comment:
Check the text between the tokens and verify that it's actually just whitespace. Otherwise if you try to join tokens from two discrete ranges the document text in the middle will get clobbered. 

Update: this is not an issue because it only operates on the target, and those would be separate targets. This means we have the reverse problem if you specify multiple targets, but this already affects existing commands, like sort. We should probably file an issue for that but it shouldn't block this change